### PR TITLE
Small refactorings

### PR DESCRIPTION
**Pull request by scan-lan**
#### the title is gonna be a laugh when you see the diff

basically, this PR addresses tickets [#103](https://zube.io/craig180885-napier/sem-group_project/c/103), [#104](https://zube.io/craig180885-napier/sem-group_project/c/104) and [#105](https://zube.io/craig180885-napier/sem-group_project/c/105), as well as: adding integration tests to fully cover (not fully test) the DAO executeStatement methods; unit test the new getWhereCondition method; and added integration tests for allCountriesIn.

i don't even remember doing that last one to be honest. sorry about the size of it, the split view in github may help.

### CODE REVIEW

File: src/main/java/com/napier/sem/DAO.java [hunk 1]
```diff
@@ -15,16 +15,39 @@
...
+                whereCondition = "country.name = '" + areaName + "'\n";
+                break;
+            case App.DISTRICT:
+                whereCondition = "city.district = '" + areaName + "'\n";
+                break;
+            default:
+                return null;
+        }
+        return whereCondition;
+    }
```
Comment (scan-lan) [inline src/main/java/com/napier/sem/DAO.java:49]: my magnum opus in this project so far

File: src/main/java/com/napier/sem/DAO.java [hunk 2]
```diff
@@ -122,29 +146,16 @@ public DAO(Connection connection)
...
-                whereClause = "WHERE country.continent = '" + areaName + "'\n";
-                break;
-            case REGION:
-                whereClause = "WHERE country.region = '" + areaName + "'\n";
-                break;
-            default:
-                whereClause = "";
-                break;
-        }
+        String whereCondition = getWhereCondition(area, areaName);
```
Comment (scan-lan) [inline src/main/java/com/napier/sem/DAO.java:149]: just look at those lines disappear

File: src/main/java/com/napier/sem/DAO.java [hunk 3]
```diff
@@ -229,69 +208,39 @@ public DAO(Connection connection)
...
 
         // Define the SQL query as a string
         String statementString = "SELECT city.name, district, city.population, country.name AS country\n" +
                 "FROM city\n" +
                 "    JOIN country ON city.countrycode = country.code\n" +
                 "WHERE city.population > 0 \n" +
-                whereClause +
+                "AND " + whereCondition +
```
Comment (scan-lan) [inline src/main/java/com/napier/sem/DAO.java:221]: notice how it works whether a `WHERE` or an `AND` is in front? sublime

File: src/main/java/com/napier/sem/App.java [hunk 4]
```diff
@@ -76,20 +76,20 @@ public static void main(String[] args) {
 
         // Use-case 5.1-5.3
         // Produce a report on all capital cities in the world organised by largest population to smallest
-         ArrayList<CapitalCity> capitalCities = dao.allCapitalCitiesIn(WORLD, "");
+//         ArrayList<CapitalCity> capitalCities = dao.allCapitalCitiesIn(WORLD, "");
```
Comment (scan-lan) [inline src/main/java/com/napier/sem/App.java:79]: this isn't pretty, i'd prefer all comments be in line. it was a mistake, there are a few other styling errors in this file. if i make any other adjustments as a result of comments i'll fix it.

File: src/main/java/com/napier/sem/DAO.java [hunk 5]
```diff
@@ -15,14 +15,40 @@
 {
     // Private properties
     private final Connection connection;
-    private static final String CONTINENT = "continent";
-    private static final String REGION = "region";
-    private static final String COUNTRY = "country";
-    private static final String DISTRICT = "district";
 
-    public DAO(Connection connection)
+    public DAO(Connection connection) { this.connection = connection; }
```
Comment (scan-lan) [inline src/main/java/com/napier/sem/DAO.java:19]: made this a one-liner so it's nicer to look at

File: src/main/java/com/napier/sem/DAO.java [hunk 6]
```diff
@@ -15,14 +15,40 @@
...
+            case App.COUNTRY:
+                whereCondition = "country.name = '" + areaName + "'\n";
+                break;
+            case App.DISTRICT:
+                whereCondition = "city.district = '" + areaName + "'\n";
+                break;
+            default:
+                return null;
+        }
+        return whereCondition;
```
Comment (scan-lan) [inline src/main/java/com/napier/sem/DAO.java:51]: my magnum opus in this project so far

File: src/main/java/com/napier/sem/DAO.java [hunk 7]
```diff
@@ -15,14 +15,40 @@
...
-    private static final String DISTRICT = "district";
 
-    public DAO(Connection connection)
+    public DAO(Connection connection) { this.connection = connection; }
+
+    /**
+     * Generates a valid condition for use with our SQL queries
+     * @param areaFilter The geographical category by which we'll be filtering
+     * @param areaName The name of the location which will be used
+     * @return The where condition string
```
Comment (scan-lan) [inline src/main/java/com/napier/sem/DAO.java:25]: this comment might be shite, open to suggestions

File: src/main/java/com/napier/sem/DAO.java [hunk 8]
```diff
@@ -120,31 +147,18 @@ public DAO(Connection connection)
...
-            case CONTINENT:
-                whereClause = "WHERE country.continent = '" + areaName + "'\n";
-                break;
-            case REGION:
-                whereClause = "WHERE country.region = '" + areaName + "'\n";
-                break;
-            default:
-                whereClause = "";
-                break;
-        }
```
Comment (scan-lan) [inline src/main/java/com/napier/sem/DAO.java:138]: ## just look at those lines disappear

File: src/main/java/com/napier/sem/App.java [hunk 9]
```diff
@@ -23,19 +23,19 @@ public static void main(String[] args) {
         // Connect to database
         if (args.length < 1)
         {
-            connection = connect("localhost:3306");
+            connection = connect("localhost:33061", "com.mysql.cj.jdbc.Driver");
```
Comment (scan-lan) [inline src/main/java/com/napier/sem/App.java:26]: `"com.mysql.cj.jdbc.Driver"` should probably be a constant, happy to change this if we want

**Review by LukasP23 — APPROVED** [review]
Tantalising

**Pull request merged**